src/python/pants/invalidation/cache_manager.py

```diff
@@ -168,42 +184,46 @@ def _results_dir_path(self, root_dir, key, stable):
-    """Ensures that a results_dir exists under the given root_dir for this versioned target.
+  def create_results_dir(self, root_dir):
+    """Ensures that a cleaned results_dir exists for invalid versioned targets.
 
-    If incremental=True, attempts to clone the results_dir for the previous version of this target
-    to the new results dir. Otherwise, simply ensures that the results dir exists.
+    Only guarantees results_dirs for invalid VTs.
     """
-    # Generate unique and stable directory paths for this cache key.
-    current_dir = self._results_dir_path(root_dir, self.cache_key, stable=False)
-    self._current_results_dir = current_dir
-    stable_dir = self._results_dir_path(root_dir, self.cache_key, stable=True)
-    self._results_dir = stable_dir
-    if self.valid:
-      # If the target is valid, both directories can be assumed to exist.
-      return
-
-    # Clone from the previous results_dir if incremental, or initialize.
-    previous_dir = self._use_previous_dir(allow_incremental, root_dir, current_dir)
-    if previous_dir is not None:
-      self.is_incremental = True
-      self._previous_results_dir = previous_dir
-      shutil.copytree(previous_dir, current_dir)
-    else:
-      safe_mkdir(current_dir)
+    self._current_results_dir = self._results_dir_path(root_dir, self.cache_key, stable=False)
+    self._results_dir = self._results_dir_path(root_dir, self.cache_key, stable=True)
 
-    # Finally, create the stable symlink.
-    relative_symlink(current_dir, stable_dir)
+    # If the target is valid, both directories are assumed to exist.
+    if not self.valid:
+      # If the vt is invalid, clean the unique file path and create a symlink to be the stable path.
+      safe_mkdir(self._current_results_dir, clean=True)
+      relative_symlink(self._current_results_dir, self._results_dir)
 
-  def _use_previous_dir(self, allow_incremental, root_dir, current_dir):
+  def copy_previous_results(self, root_dir):
+    """Use the latest valid results_dir as the starting contents of the current results_dir.
+
+    Should be called after the cache is checked, since previous_results are not useful if there is a cached artifact.
+    """
+    # TODO(mateo): An immediate followup removes the root_dir param, it is identical to the task.workdir.
+    if not self.previous_cache_key:
+      return None
+    previous_path = self._results_dir_path(root_dir, self.previous_cache_key, stable=False)
+    if os.path.isdir(previous_path):
+      self.is_incremental = True
+      safe_rmtree(self._current_results_dir)
+      shutil.copytree(previous_path, self._current_results_dir)
+    safe_mkdir(self._current_results_dir)
+    relative_symlink(self._current_results_dir, self.results_dir)
+    # Set the self._previous last, so that it is only True after the copy completed.
+    self._previous_results_dir = previous_path
+
+  def _use_previous_dir(self, allow_incremental, root_dir):
```

Comment:
Yep, thanks for the catch. All the functions of this method are covered elsewhere, so will delete.

It was a miss when splitting from the followup commit, which cleans up the cache_manager's more mutation-heavy choices: https://github.com/mateor/pants/commit/62e49a3f5b5ce2a6814180ec0bff0cd94c56f957

It needs to be updated but I intend to put up a roughly analogous RB as a followup.